src/allocation_engine/interval_tree.rs

```diff
@@ -33,14 +37,139 @@ pub struct Constraint {
+    /// ```
+    pub fn align_forward(&self, align: u64) -> Result<Range> {
+        match align {
+            0 | 1 => Ok(*self),
+            _ => {
+                // Check alignment to be a power of two.
+                if !align.is_power_of_two() {
+                    return Err(Error::InvalidAlignment);
+                }
+                // !(align - 1)) will set all bits starting with the one already
+                // set in the alignment value, all bits until that one will be 0.
+                //
+                // (self.range_start + align - 1) & !(align - 1) will round up the lower
+                // bound of this interval to the first multiple of `align`. If
+                // the aligned address does not belong to this interval, we will
+                // return error.
+                if let Some(range_start) = self
+                    .range_start
+                    .checked_add(align - 1)
+                    .map(|v| v & !(align - 1))
+                {
+                    if range_start <= self.range_end {
+                        return Range::new(range_start, self.range_end);
+                    }
+                }
+                Err(Error::UnalignedAddress)
+            }
+        }
+    }
+
+    /// Create a new Range object with range_start being the last value from the
+    /// range that is aligned to `align`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// extern crate vm_allocator;
+    /// use vm_allocator::allocation_engine::Range;
+    /// use vm_allocator::Result;
+    ///
+    /// fn intervals_align_back() -> Result<Range> {
+    ///     let range_to_align = Range::new(2, 6)?;
+    ///     let res = range_to_align.align_back(4, 1)?;
+    ///     assert_eq!(res.length(), 1);
+    ///     Ok(res)
+    /// }
+    ///
+    /// intervals_align_back().unwrap();
+    pub fn align_back(&self, align: u64, size: u64) -> Result<Range> {
+        // This method is used with `AllocPolicy::LastMatch`. It is used to
+        // find the last value from a range that is aligned to `align`.
+        // First step is to subtract from the end of the range the size of the
+        // desired memory slot.
+        let mut candidate_address = self
+            .range_end
+            .checked_sub(size)
+            .ok_or(Error::UnalignedAddress)?
+            + 1;
+
+        // Return `Error::InvalidAlignment` if `align` is not a power of two.
+        if !align.is_power_of_two() {
+            return Err(Error::InvalidAlignment);
+        }
+
+        // The desired memory slot is represented by a range that starts at the
+        // highest multiple of align that is less than or equal to `candidate_address`.
+        // The highest multiple of n less than or equal to b is given by the
+        // equation floor(b/n) * n. If the result is less than the lower bound
+        // of the range, then no such multiple exists.
+        candidate_address = (candidate_address / align) * align;
```

Comment:
This operation is inefficient. There is no reason to use division, and we can compute the aligned address with bits operations similar to how it's done for the align_up.